aesd-char-driver/main.c

```diff
@@ -0,0 +1,339 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes, edited by Guruprashanth Krishnakumar
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h> //size_t    
+#include <linux/cdev.h>
+#include <linux/kernel.h> //containerof
+#include <linux/slab.h> //kmalloc
+#include <linux/uaccess.h>	/* copy_*_user */
+#include <linux/fs.h> // file_operations
+
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Guruprashanth Krishnakumar"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+/*
+*   Allocated or reallocate the requested amount of memory
+*
+*   Args:
+*       ptr - store pointer to requested memory here
+*       size - allocated size bytes of memory
+*   Params:
+*       0 if success, -ENOMEM if failed
+*/
+static int allocate_memory(struct working_buffer *ptr, int size);
+
+static int allocate_memory(struct working_buffer *ptr, int size)
+{
+    int retval = 0;
+    char *temp_ptr;
+    if(ptr->size == 0)
+    {
+        ptr->buffer = kmalloc(size,GFP_KERNEL);
+        if(!ptr->buffer)
+        {
+            retval = -ENOMEM;
+            //Add a goto statement. We need to free COPY_BUFFER and 
+            //UNLOCK MUTEX   
+        }
+    }
+    else
+    {
+        temp_ptr = krealloc(ptr->buffer,ptr->size + size,GFP_KERNEL);
+        if(!temp_ptr)
+        {
+            kfree(ptr->buffer);
+            retval = -ENOMEM;
+            //Add a goto statement. We need to free COPY_BUFFER and WORKING_BUFFER
+            //UNLOCK MUTEX   
+            
+        }
+        else
+        {
+            ptr->buffer = temp_ptr;
+        }
+    }
+    return retval;
+}
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *info_dev;
+    PDEBUG("open");
+    //find the aesd_dev structure associated with the passed inode and assign it to private_data
+    info_dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = info_dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * Remove assignment of private data.
+     */
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *data;
+    size_t relative_byte_offset;
+    struct aesd_buffer_entry *ret_buf;
+    int bytes_to_eob;
+    int bytes_to_copy;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    if(!filp || !filp->private_data)
```

Comment:
@GuruprashanthKrishnakumar good defensive programming here but I think you can assume `filp` is setup for you by the kernel fops functions and not null.  See https://github.com/cu-ecen-aeld/ldd3/blob/master/scull/main.c#L339 as a reference example for kernel ops parameter assumptions.